src/library/browse/browsethread.cpp

```diff
@@ -14,80 +14,109 @@ namespace {
 constexpr int kRowBatchSize = 10;
 } // namespace
 
-QWeakPointer<BrowseThread> BrowseThread::m_weakInstanceRef;
-static QMutex s_Mutex;
-
 /*
  * This class is a singleton and represents a thread
  * that is used to read ID3 metadata
  * from a particular folder.
  *
  * The BrowseTableModel uses this class.
- * Note: Don't call getInstance() from places
+ * Note: Don't call getInstanceRef() from places
  * other than the GUI thread. BrowseThreads emit
  * signals to BrowseModel objects. It does not
  * make sense to use this class in non-GUI threads
  */
-BrowseThread::BrowseThread(QObject *parent)
-        : QThread(parent) {
-    m_bStopThread = false;
-    m_model_observer = nullptr;
-    //start Thread
+BrowseThread::BrowseThread(QObject* parent)
+        : QThread{parent}, m_modelObserver{}, m_bRun{} {
+    // Start thread
     start(QThread::LowPriority);
-
+    qDebug() << "Wait to start browser background thread";
+    // Wait until the thread is running
+    m_mutex.lock();
+    while (!m_bRun) {
+        m_condition.wait(&m_mutex);
+    }
```

Comment:
I have not understood the loop. Isn't there only one event/codition we are waiting for? m_bRun shall become a semaphore? On the other hand syncronizing the thread start by halting the main thread feels wrong. It would be better to protect the shared sources explicit. What are the shared resources in this case we have tsan complains?

Reply:
Yes, we are waiting for m_bRun to be set to true, as set by the thread::run method. A semaphore would work as well, but since we already have the condition, I take advantage of it.

Reply:
> A semaphore would work as well, but since we already have the condition, I take advantage of it.

This reusing makees the code harder to read. So I would prefer two primitived for two tasks. 

The ThreadSanitizer warning is really confusing, what does it check for? Whether a variable is written/read from two threads? It must be a control structure in Qt. However waiting for the thread to start should not change the issue for my undertsanding. 

But if it is the solution i don't mind.

Reply:
How does the solution work? I am afraid instead of a solution, we trick ThreadSanitizer.
Can, it happen that it treated the unrelated mutex as a proper guard?
Qt allows to send signals and reconnect them concurrently. So it should not be a problem. 
Maybe only a missing atomic annotation in Qt? 

Is there a way to find out what 0x0001466d5f28 is and which mutexted I are considered? 

Or is it s QT bug? Is 2.4 with Qt 5 affected as well.

Reply:
Got it, I n this thread they discussed exactly our problem. Solution is to build Qt wir sanitizer flag: 
https://lists.qt-project.org/pipermail/interest/2022-November/038561.html

Conclusion here: We can safely ignore the issue.

Reply:
I have build Qt with the sanitizer flag, though I used the normal way (passing the -fsanitize=thread flag), so unless this configure option does something else (I doubt it), the problem is still there.

I don't see you problem with using one and the same mutex and condition.

Reply:
Anyway, I will see if I can build qt in my vcpkg with this configure option, and if that makes a difference. But I would appreciate it if you can point me to where I need to change this, given my inexperience with vcpkg.

Reply:
I think you can add it here: 
https://github.com/mixxxdj/vcpkg/blob/74ff6c7ceb657e9f099626b5834884184209bf50/ports/qt5-base/portfile.cmake#L94

Reply:
Ok, I will give that a try.

Reply:
I am building against Qt6 though

Reply:
which seems to use a different approach... Any idea how to do it there?

Reply:
Anyway, I am now building Qt5. But it is on purpose that the 2.5 branch of mixxx vcpkg has Qt6 as the default? I am confused now as to what we are supposed to use for Mixxx 2.5.

Reply:
I am running into lots of problems trying to build against Qt5, so it would be much appreciated if you could help me building Qt6 with vcpkg with this Qt sanitize thread option.

Reply:
Mixxx 2.4.2 will be our last Qt5 release. So there is IMHO no need to deal with QT5 in this case. We just keep the Qt5 support in our 2.5 branch just in case we want to test regresions. 

I suggest to issue a PR against our vcpkg branch and use the build package downloaded from GitHub. For my undersanding you jsut need to add the sanitizer flag at the position I pointed out above.

Reply:
Yes, I prefer to use Qt 6, but you pointed it out in the qt5 portfile.cmake. The qt6 portfile.cmake is completely different, and I don't see where to add this option.

Reply:
It looks like this works. I'd prefer to set this from my triplets file arm64-osx-tsan.cmake, but for now this have to do.

diff --git a/ports/qtbase/portfile.cmake b/ports/qtbase/portfile.cmake
index 8b9e3b0d3..4a96412d7 100644
--- a/ports/qtbase/portfile.cmake
+++ b/ports/qtbase/portfile.cmake
@@ -92,7 +92,7 @@ INVERTED_FEATURES
     )
 
 list(APPEND FEATURE_OPTIONS -DCMAKE_DISABLE_FIND_PACKAGE_Libudev:BOOL=ON)
-list(APPEND FEATURE_OPTIONS -DFEATURE_xml:BOOL=ON)
+list(APPEND FEATURE_OPTIONS -DFEATURE_xml:BOOL=ON -DFEATURE_sanitize_thread:BOOL=ON)